src/main/java/com/google/step/servlets/HomePageServlet.java

```diff
@@ -0,0 +1,208 @@
+    List<List<Map<String, Object>>> homePageDealsMapList = new ArrayList<>();
+    for (List<Deal> dealList : homePageDeals) {
+      List<Map<String, Object>> homePageSectionDealMaps = new ArrayList<>();
+      for (Deal deal : dealList) {
+        User user = userManager.readUser(deal.posterId);
+        Restaurant restaurant = restaurantManager.readRestaurant(deal.restaurantId);
+        List<Tag> tags = tagManager.readTags(dealTagManager.getTagIdsOfDeal(deal.id));
+        int votes = voteManager.getVotes(deal.id);
+        Map<String, Object> homePageDealMap =
+            JsonFormatter.getBriefHomePageDealMap(deal, user, restaurant, tags, votes);
+        homePageSectionDealMaps.add(homePageDealMap);
+      }
+      homePageDealsMapList.add(homePageSectionDealMaps);
+    }
+    return homePageDealsMapList;
+  }
+
+  /** Retrieves deals posted by tags followed by user */
+  private List<Deal> getDealsPublishedByTags(List<Long> tagIds) {
+    List<Long> dealIdResults = new ArrayList<>();
+    for (Long id : tagIds) {
+      List<Long> dealIdsWithTag = dealTagManager.getDealIdsWithTag(id);
+      dealIdResults.addAll(dealIdsWithTag);
+    }
+    // Get rid of duplicate dealID (Deals with multiple tags)
+    List<Long> dealsWithoutDuplicates = new ArrayList<>(new HashSet<>(dealIdResults));
+    return dealManager.readDeals(dealsWithoutDuplicates);
+  }
+
+  /** Sorts deals based on new (Newest to oldest) */
+  private List<Deal> sortDealsBasedOnNew(List<Deal> deals) {
+    Collections.sort(
+        deals,
+        new Comparator<Deal>() {
+          @Override
+          public int compare(Deal deal1, Deal deal2) { // TODO fix parse
+            return LocalDateTime.parse(deal2.creationTimeStamp)
+                .compareTo(LocalDateTime.parse(deal1.creationTimeStamp)); // Descending
+          }
+        });
+    return deals;
+  }
+
+  /** Sorts deals based on hot score (Highest to lowest) */
+  private double epochSeconds(String timestamp) {
+    LocalDateTime time = LocalDateTime.parse(timestamp);
+    long epoch = time.atZone(ZoneId.of(LOCATION)).toEpochSecond();
+    return epoch;
+  }
+
+  /**
+   * Calculates a hot score for each deal entity, which takes into account both the time and the
+   * amount of votes it got
+   */
+  private double calculateHotScore(Deal deal) {
+    int netVotes = voteManager.getVotes(deal.id);
+    double order = Math.log(Math.max(Math.abs(netVotes), 1));
+    int sign = 0;
+    if (netVotes > 0) sign = 1;
+    else if (netVotes < 0) sign = -1;
+    double seconds = epochSeconds((String) deal.creationTimeStamp) - OLDEST_DEAL_TIMESTAMP;
+    return sign * order + seconds / 45000;
+  }
+
+  /** Sorts deals based on votes (Highest to lowest) */
+  private List<Deal> getSortedDealsBasedOnValue(List<Deal> deals, String attribute) {
+    List<DealPair> dealPairs = new ArrayList<>();
+    for (Deal deal : deals) {
+      if (attribute.equals("hotScore")) {
```

Comment:
I don't really like this if else statements with hardcoded strings. Can you split the function into `getSortedDealsByHotScore` and `getSortedDealsByVotes`?

Reply:
also try not to call getVotes so many times, because getVotes is not O(1). Preferably, call it only once for each dealId.

Reply:
i think its only called once for each deal though, to create the deal pair

Reply:
it is called in `getHomePageDealMaps`, in `calculateHotScore`, and in `sortDealsBasedOnVotes`. If it is too messy to modify it to pass vote counters everywhere, then you can just leave it.